src/main/java/edu/cornell/kfs/fp/service/impl/RecurringDisbursementVoucherPaymentMaintenanceServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+        if (ObjectUtils.isNotNull(paymentGroup)) {
+            String paymentStatusCode = paymentGroup.getPaymentStatus().getCode();
+            checkCancelPermissionAndError(user, paymentStatusCode);
+            if (!(PdpConstants.PaymentStatusCodes.CANCEL_PAYMENT.equals(paymentStatusCode))) {
+                return cancelPaymentDetail(paymentDetailId, note, user, paymentGroup, paymentStatusCode);
+            } else {
+                LOG.debug("cancelPendingPayment() Pending payment group has already been cancelled; exit method.");
+                return true;
+            }
+        } else {
+            LOG.debug("cancelPendingPayment() Pending payment not found.");
+            GlobalVariables.getMessageMap().putError(KFSConstants.GLOBAL_ERRORS, PdpKeyConstants.PaymentDetail.ErrorMessages.ERROR_PAYMENT_NOT_FOUND);
+            return false;
+        }
+    }
+
+    private void checkCancelPermissionAndError(Person user, String paymentStatusCode) {
+        if (!hasCancelPermission(user)) {
+            LOG.warn("checkCancelPermissionAndError() Payment status is " + paymentStatusCode + "; user does not have rights to cancel. This should not happen.");
+            throw new RuntimeException("checkCancelPermissionAndError() Payment status is " + paymentStatusCode + "; user does not have rights to cancel. This should not happen.");
+        }
+    }
+
+    private boolean cancelPaymentDetail(Integer paymentDetailId, String noteText, Person user, PaymentGroup paymentGroup, String paymentStatus) {
+        LOG.debug("cancelPaymentDetail() Payment status is " + paymentStatus + "; continue with cancel.");
+        if (PdpConstants.PaymentStatusCodes.OPEN.equals(paymentStatus) || PdpConstants.PaymentStatusCodes.HELD_CD.equals(paymentStatus)) {
+            changePaymentStatusToCancelAndAddUpdatePaymentDetail(paymentDetailId, noteText, user, paymentGroup, paymentStatus);
+        } else {
+            LOG.error("cancelPaymentDetail(): Recurring DVs should only cancel OPEN payments, we found a status of " + paymentStatus);
+            GlobalVariables.getMessageMap().putError(KFSConstants.GLOBAL_ERRORS, PdpKeyConstants.PaymentDetail.ErrorMessages.ERROR_PAYMENT_INVALID_STATUS_TO_CANCEL);
+            return false;
+        }
+        return true;
+    }
+    
+    private void changePaymentStatusToCancelAndAddUpdatePaymentDetail(Integer paymentDetailId, String noteText, Person user, PaymentGroup paymentGroup, String paymentStatus) {
+        changeStatus(paymentGroup, PdpConstants.PaymentStatusCodes.CANCEL_PAYMENT, PdpConstants.PaymentChangeCodes.CANCEL_PAYMENT_CHNG_CD, noteText, user);
+        updatePaymentDetail(paymentDetailId, noteText);
+        LOG.debug("changePaymentStatusToCancelAndAddUpdatePaymentDetail() Pending payment cancelled; exit method.");
+    }
+
+    private void updatePaymentDetail(Integer paymentDetailId, String noteText) {
+        Map primaryKeys = new HashMap();
```

Comment:
Unless the BO service is unable to accept wildcard-related Maps, I would suggest using generics here.